File: patient-ios-app/MediHubPatient/Core/Auth/KeycloakAuthService.swift
import AppAuth
import Foundation
import UIKit

/// Keycloak / OIDC auth façade backed by AppAuth-iOS (KC-3).
///
/// Additive — does not replace `AuthManager`. While
/// `FeatureFlags.keycloakSsoEnabled` is false (the default until prerequisite
/// P-2 lands) nothing in the app uses this class.
@MainActor
final class KeycloakAuthService: ObservableObject {
    static let shared = KeycloakAuthService()

    /// Held during the in-flight authorization so AppAuth can forward the
    /// redirect callback from `MediHubPatientApp.onOpenURL`.
    private var currentAuthFlow: OIDExternalUserAgentSession?

    /// Last successful auth state (also persisted to Keychain).
    @Published private(set) var authState: OIDAuthState?

    private init() {
        authState = Self.loadAuthState()
    }

    // MARK: - Public API

    var isConfigured: Bool { KeycloakConfig.isConfigured }

    /// True when a Keycloak session is active (used by APIClient).
    var hasActiveSession: Bool { authState?.isAuthorized == true }

    /// Discover OIDC endpoints, then run the Authorization Code + PKCE flow
    /// via `SFSafariViewController` / `ASWebAuthenticationSession`.
    func login(presenting viewController: UIViewController) async throws {
        guard isConfigured else { throw KeycloakError.notConfigured }
        guard let issuerURL = URL(string: KeycloakConfig.issuer),
              let redirectURL = URL(string: KeycloakConfig.redirectURI)
        else { throw KeycloakError.invalidConfiguration }

        let config = try await discoverConfiguration(issuer: issuerURL)
        let request = OIDAuthorizationRequest(
            configuration: config,
            clientId: KeycloakConfig.clientID,
            clientSecret: nil,
            scopes: [OIDScopeOpenID, OIDScopeProfile, OIDScopeEmail, "offline_access"],
            redirectURL: redirectURL,
            responseType: OIDResponseTypeCode,
            additionalParameters: nil
        )

        let state = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<OIDAuthState, Error>) in
            currentAuthFlow = OIDAuthState.authState(
                byPresenting: request,
                presenting: viewController
            ) { authState, error in
                if let authState {
                    continuation.resume(returning: authState)
                } else {
                    continuation.resume(throwing: error ?? KeycloakError.unknown)
                }
            }
        }

        currentAuthFlow = nil
        authState = state
        persist(state: state)
    }

    /// Forward a redirect URL from the app delegate / SwiftUI `onOpenURL`.
    @discardableResult
    func resume(url: URL) -> Bool {
        guard let flow = currentAuthFlow else { return false }
        let handled = flow.resumeExternalUserAgentFlow(with: url)
        if handled { currentAuthFlow = nil }
        return handled
    }

    /// Returns a non-expired access token (refreshing via the refresh token
    /// when necessary). Returns `nil` when no OIDC session is active.
    func freshAccessToken() async throws -> String? {
        guard let state = authState else { return nil }
        return try await withCheckedThrowingContinuation { continuation in
            state.performAction { accessToken, _, error in
                if let error {
                    continuation.resume(throwing: error)
                } else {
                    // Persist updated tokens (refresh may have rotated them).
                    Task { @MainActor in self.persist(state: state) }
                    continuation.resume(returning: accessToken)
                }
            }
        }
    }

    /// Clear the persisted OIDC session (matches `AuthManager.logout`).
    func clear() {
        authState = nil
        KeychainHelper.shared.clearOidc()
    }

    // MARK: - Private

    private func discoverConfiguration(issuer: URL) async throws -> OIDServiceConfiguration {
        try await withCheckedThrowingContinuation { continuation in
            OIDAuthorizationService.discoverConfiguration(forIssuer: issuer) { config, error in
                if let config {
                    continuation.resume(returning: config)
                } else {
                    continuation.resume(throwing: error ?? KeycloakError.discoveryFailed)
                }
            }
        }
    }

    private func persist(state: OIDAuthState) {
        let data: Data
        if #available(iOS 12.0, *) {
            data = (try? NSKeyedArchiver.archivedData(
                withRootObject: state,
                requiringSecureCoding: true
            )) ?? Data()
        } else {
            data = NSKeyedArchiver.archivedData(withRootObject: state)
        }
        KeychainHelper.shared.oidcAuthState = data
        KeychainHelper.shared.oidcAccessToken = state.lastTokenResponse?.accessToken
        KeychainHelper.shared.oidcIdToken = state.lastTokenResponse?.idToken
    }

    private static func loadAuthState() -> OIDAuthState? {
        guard let data = KeychainHelper.shared.oidcAuthState else { return nil }
        if #available(iOS 12.0, *) {
            return try? NSKeyedUnarchiver.unarchivedObject(
                ofClass: OIDAuthState.self,
                from: data
            )
        } else {
            return NSKeyedUnarchiver.unarchiveObject(with: data) as? OIDAuthState
        }
    }
}

// MARK: - Errors

enum KeycloakError: LocalizedError {
    case notConfigured
    case invalidConfiguration
    case discoveryFailed
    case unknown

    var errorDescription: String? {
        switch self {
        case .notConfigured: "Keycloak SSO is not configured for this build."
        case .invalidConfiguration: "Keycloak issuer/redirect URI is invalid."
        case .discoveryFailed: "Unable to reach the Keycloak discovery endpoint."
        case .unknown: "SSO login failed."
        }
    }
}

#if DEBUG
extension KeycloakAuthService {
    /// Debug-only bypass for the `ASWebAuthenticationSession` browser sheet.
    ///
    /// XCUITest cannot reliably automate `ASWebAuthenticationSession` (Apple
    /// runs it out-of-process and sandboxes it from the test runner), so we
    /// expose this entry point so test harnesses — and developers running a
    /// build with `MEDIHUB_KEYCLOAK_E2E_BYPASS=1` — can exercise the
    /// post-redirect code path (Keychain persistence, `hasActiveSession`,
    /// API client wiring) without driving the system browser. See
    /// `docs/keycloak-implementation-gaps.md` §4 (G-5) for the rationale.
    ///
    /// Synthesizes the same `OIDAuthState` shape AppAuth produces after a
    /// successful redirect, then routes through the existing `persist`
    /// path so the test exercises the real keychain flow.
    func acceptDebugSession(
        accessToken: String,
        refreshToken: String? = nil,
        idToken: String? = nil,
        expiresIn: TimeInterval = 3600
    ) {
        // Fall back to a docker-compose default when the build is unconfigured
        // (test harnesses, CI without scheme env vars). Never crash the host
        // process — assertion-fail and bail so the caller sees a no-op.
        let issuerString = KeycloakConfig.isConfigured
            ? KeycloakConfig.issuer
            : "http://localhost:8081/realms/hms"
        let redirectString = KeycloakConfig.redirectURI.isEmpty
            ? "com.bitnesttechs.hms.patient.native:/oauth2redirect"
            : KeycloakConfig.redirectURI
        let clientId = KeycloakConfig.clientID.isEmpty
            ? "hms-patient-ios"
            : KeycloakConfig.clientID

        guard
            let authEndpoint = URL(string: "\(issuerString)/protocol/openid-connect/auth"),
            let tokenEndpoint = URL(string: "\(issuerString)/protocol/openid-connect/token"),
            let redirectURL = URL(string: redirectString)
        else {
            assertionFailure("acceptDebugSession: KeycloakConfig produced an unparseable URL (issuer=\(issuerString), redirect=\(redirectString)).")
            return
        }
        let serviceConfig = OIDServiceConfiguration(
            authorizationEndpoint: authEndpoint,
            tokenEndpoint: tokenEndpoint
        )

        let authRequest = OIDAuthorizationRequest(
            configuration: serviceConfig,
            clientId: clientId,
            clientSecret: nil,
            scopes: [OIDScopeOpenID, OIDScopeProfile, OIDScopeEmail],
            redirectURL: redirectURL,
            responseType: OIDResponseTypeCode,
            additionalParameters: nil
        )

        let authResponseParams: [String: NSString] = [
            "code": "debug-auth-code" as NSString,
            "state": (authRequest.state ?? "debug-state") as NSString,
        ]
        let authResponse = OIDAuthorizationResponse(
            request: authRequest,
            parameters: authResponseParams
        )

        guard let tokenRequest = authResponse.tokenExchangeRequest() else {
            assertionFailure("AppAuth refused to build a token-exchange request from the synthetic auth response.")
            return
        }

        var tokenParams: [String: NSString] = [
            "access_token": accessToken as NSString,
            "token_type": "Bearer" as NSString,
            "expires_in": String(Int(expiresIn)) as NSString,
            "scope": "openid profile email" as NSString,
        ]
        if let refreshToken {
            tokenParams["refresh_token"] = refreshToken as NSString
        }
        if let idToken {
            tokenParams["id_token"] = idToken as NSString
        }
        let tokenResponse = OIDTokenResponse(request: tokenRequest, parameters: tokenParams)

        let state = OIDAuthState(
            authorizationResponse: authResponse,
            tokenResponse: tokenResponse
        )
        authState = state
        persist(state: state)
    }
}
#endif

File: patient-android-app/app/src/main/java/com/bitnesttechs/hms/patient/core/auth/KeycloakAuthService.kt
package com.bitnesttechs.hms.patient.core.auth

import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.net.Uri
import com.bitnesttechs.hms.patient.BuildConfig
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.suspendCancellableCoroutine
import net.openid.appauth.AuthState
import net.openid.appauth.AuthorizationException
import net.openid.appauth.AuthorizationRequest
import net.openid.appauth.AuthorizationResponse
import net.openid.appauth.AuthorizationService
import net.openid.appauth.AuthorizationServiceConfiguration
import net.openid.appauth.EndSessionRequest
import net.openid.appauth.ResponseTypeValues
import net.openid.appauth.TokenResponse
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.resume

/**
 * Keycloak / OIDC auth façade backed by AppAuth-Android (KC-3).
 *
 * This service is additive — it does not replace [AuthRepository]. While
 * `FeatureFlagManager.keycloakSsoEnabled` is false (the default until
 * prerequisite P-2 lands), nothing in the app uses this class.
 *
 * Lifecycle:
 *  - [buildAuthorizationIntent] returns an Intent that the Activity launches
 *    through the AppAuth `ActivityResultContract` / Custom Tabs flow.
 *  - [handleAuthorizationResponse] consumes the result Intent and completes
 *    the authorization code + PKCE token exchange.
 *  - [freshAccessToken] returns a (possibly refreshed) access token for the
 *    OkHttp interceptor.
 *  - [buildEndSessionIntent] assembles an RP-initiated logout Intent.
 */
@Singleton
class KeycloakAuthService @Inject constructor(
    @ApplicationContext private val context: Context,
    private val tokenStorage: TokenStorage
) {
    private val authorizationService: AuthorizationService by lazy {
        AuthorizationService(context)
    }

    private val redirectUri: Uri = Uri.parse(BuildConfig.KEYCLOAK_REDIRECT_URI)

    /** True when the build has a non-empty issuer configured. */
    val isConfigured: Boolean
        get() = BuildConfig.KEYCLOAK_ISSUER.isNotBlank()

    /**
     * Discover the OIDC endpoints for the configured issuer.
     *
     * @throws IllegalStateException if `BuildConfig.KEYCLOAK_ISSUER` is blank.
     */
    suspend fun fetchConfiguration(): AuthorizationServiceConfiguration {
        check(isConfigured) {
            "Keycloak issuer is not configured (BuildConfig.KEYCLOAK_ISSUER is empty)"
        }
        val issuerUri = Uri.parse(BuildConfig.KEYCLOAK_ISSUER)
        return suspendCancellableCoroutine { cont ->
            AuthorizationServiceConfiguration.fetchFromIssuer(issuerUri) { config, ex ->
                if (config != null) {
                    cont.resume(config)
                } else {
                    cont.resumeWith(
                        Result.failure(ex ?: IllegalStateException("Unknown discovery error"))
                    )
                }
            }
        }
    }

    /**
     * Build an authorization (login) Intent for the Activity to launch.
     * Uses Auth Code + PKCE.
     */
    suspend fun buildAuthorizationIntent(): Intent {
        val config = fetchConfiguration()
        val request = AuthorizationRequest.Builder(
            config,
            BuildConfig.KEYCLOAK_CLIENT_ID,
            ResponseTypeValues.CODE,
            redirectUri
        )
            .setScope("openid profile email offline_access")
            .build()

        // Persist a stub AuthState tied to this config so later calls can reuse it
        saveAuthState(AuthState(config))
        return authorizationService.getAuthorizationRequestIntent(request)
    }

    /**
     * Exchange the authorization response for tokens and persist the resulting
     * [AuthState]. Returns the completed AuthState on success.
     */
    suspend fun handleAuthorizationResponse(data: Intent): AuthState {
        val response = AuthorizationResponse.fromIntent(data)
        val ex = AuthorizationException.fromIntent(data)
        if (response == null) {
            throw ex ?: IllegalStateException("AuthorizationResponse was null")
        }

        val state = loadAuthState() ?: AuthState(response, ex)
        state.update(response, ex)

        val tokenResponse = exchangeCode(response)
        state.update(tokenResponse, null)
        saveAuthState(state)
        return state
    }

    private suspend fun exchangeCode(response: AuthorizationResponse): TokenResponse =
        suspendCancellableCoroutine { cont ->
            authorizationService.performTokenRequest(response.createTokenExchangeRequest()) { token, ex ->
                if (token != null) {
                    cont.resume(token)
                } else {
                    cont.resumeWith(
                        Result.failure(ex ?: IllegalStateException("Token exchange failed"))
                    )
                }
            }
        }

    /**
     * Returns a non-expired access token, refreshing via the refresh token when
     * necessary. Returns null when no OIDC session is active.
     */
    suspend fun freshAccessToken(): String? {
        val state = loadAuthState() ?: return null
        return suspendCancellableCoroutine { cont ->
            state.performActionWithFreshTokens(authorizationService) { token, _, ex ->
                if (ex != null) {
                    cont.resume(null)
                    return@performActionWithFreshTokens
                }
                saveAuthState(state)
                cont.resume(token)
            }
        }
    }

    /**
     * Build an RP-initiated end-session Intent for the Activity to launch.
     * Returns null when there is no OIDC session or the id token has been
     * discarded.
     */
    fun buildEndSessionIntent(postLogoutRedirect: Uri = redirectUri): Intent? {
        val state = loadAuthState() ?: return null
        val config = state.authorizationServiceConfiguration ?: return null
        val idToken = state.idToken ?: return null
        val request = EndSessionRequest.Builder(config)
            .setIdTokenHint(idToken)
            .setPostLogoutRedirectUri(postLogoutRedirect)
            .build()
        return authorizationService.getEndSessionRequestIntent(request)
    }

    fun clear() {
        tokenStorage.clearOidc()
    }

    fun dispose() {
        authorizationService.dispose()
    }

    // ------------------------------------------------------------------
    // Persistence
    // ------------------------------------------------------------------

    internal fun saveAuthState(state: AuthState) {
        tokenStorage.oidcAuthStateJson = state.jsonSerializeString()
        tokenStorage.oidcAccessToken = state.accessToken
        tokenStorage.oidcIdToken = state.idToken
    }

    internal fun loadAuthState(): AuthState? {
        val json = tokenStorage.oidcAuthStateJson ?: return null
        return try {
            AuthState.jsonDeserialize(json)
        } catch (_: org.json.JSONException) {
            null
        }
    }

    /** PendingIntent helpers for future integration with launchAuthorizationRequest(). */
    @Suppress("unused")
    internal fun pendingIntent(intent: Intent, requestCode: Int): PendingIntent =
        PendingIntent.getActivity(
            context,
            requestCode,
            intent,
            PendingIntent.FLAG_MUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
        )
}
